src/shared/syscheck_op.c

```diff
@@ -906,6 +926,207 @@ const char *get_group(__attribute__((unused)) int gid) {
+                                         &pDacl,                // Pointer to ACL
+                                         &fDaclDefaulted);      // Flag set to the value of the SE_DACL_DEFAULTED flag
+
+    if (bRtnBool == FALSE) {
+        dwErrorCode = GetLastError();
+        mwarn("GetSecurityDescriptorDacl failed. GetLastError returned: %ld", dwErrorCode);
+        os_free(pSecurityDescriptor);
+        return dwErrorCode;
+    }
+
+    // Check whether no DACL or a NULL DACL was retrieved from the security descriptor buffer.
+    if (fDaclPresent == FALSE || pDacl == NULL) {
+        mwarn("No DACL was found (all access is denied), or a NULL DACL (unrestricted access) was found.");
+        os_free(pSecurityDescriptor);
+        return -1;
+    }
+
+    // Retrieve the ACL_SIZE_INFORMATION structure to find the number of ACEs in the DACL.
+    bRtnBool = GetAclInformation(
+                                 pDacl,                 // Pointer to an ACL
+                                 &aclsizeinfo,          // Pointer to a buffer to receive the requested information
+                                 sizeof(aclsizeinfo),   // The size, in bytes, of the buffer
+                                 AclSizeInformation);   // Fill the buffer with an ACL_SIZE_INFORMATION structure
+
+    if (bRtnBool == FALSE) {
+        dwErrorCode = GetLastError();
+        mwarn("GetAclInformation failed. GetLastError returned: %ld", dwErrorCode);
+        os_free(pSecurityDescriptor);
+        return dwErrorCode;
+    }
+
+    // Loop through the ACEs to get the information.
+    for (cAce = 0; cAce < aclsizeinfo.AceCount; cAce++) {
+        // Get ACE info
+        if (GetAce(pDacl, cAce, (LPVOID*)&pAce) == FALSE) {
+            mwarn("GetAce failed. GetLastError returned: %ld", GetLastError());
+            continue;
+        }
+
+        written = copy_ace_info(pAce, permissions, perm_size);
+
+        if (written > 0) {
+            permissions += written;
+            perm_size -= written;
+
+            if (perm_size > 0) {
+                continue;
+            }
+        }
+    }
+
+    return ERROR_SUCCESS;
+}
+
+unsigned int FILETIME_to_POSIX(FILETIME ft) {
+    // takes the last modified date
+    LARGE_INTEGER date, adjust;
+    date.HighPart = ft.dwHighDateTime;
+    date.LowPart = ft.dwLowDateTime;
+
+    // 100-nanoseconds = milliseconds * 10000
+    adjust.QuadPart = 11644473600000 * 10000;
+
+    // removes the diff between 1970 and 1601
+    date.QuadPart -= adjust.QuadPart;
+
+    // converts back from 100-nanoseconds to seconds
+    return date.QuadPart / 10000000;
+}
+
+unsigned int get_registry_mtime(HKEY hndl) {
+    FILETIME lpftLastWriteTime;
+    DWORD dwRtnCode = 0;
+
+    dwRtnCode = RegQueryInfoKeyA(hndl, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, &lpftLastWriteTime);
+
+    if (dwRtnCode != ERROR_SUCCESS) {
+        mwarn("Couldn't get modification time for registry key.");
```

Comment:
In case of an error, the 'FILETIME_to_POSIX()` function cannot be called. Check that the returned value is valid in case of error.

Reply:
Done 7968a28